### PR TITLE
feat: credentials framework

### PR DESCRIPTION
**Pull request by g-linville**
You can read more about it in the doc I am adding in this PR, but basically this adds a "credentials framework" so that we have an easy way to supply secret values in environment variables to tools that need them. First you need to have a credential provider tool that prints a JSON object in the following format: `{"env":{"ENV_VAR_1": "value1", "ENV_VAR_2": "value2"}}`. Then, you can reference the credential tool in a script like this:

```
name: my-tool
credential: my-credential-tool.gpt

(tool body here)
```

The key-value pairs of environment variables that `my-credential-tool.gpt` returns will be added to the environment for the main tool's execution.

### CODE REVIEW

File: pkg/engine/engine.go
Comment (ibuildthecloud) [inline pkg/engine/engine.go:215]: Why do we give the credential tools to the LLM?

File: pkg/credentials/credential.go
Comment (ibuildthecloud) [inline pkg/credentials/credential.go:10]: Should this be toolname or tool? Not id.  FYI, ID's are not really "portable" They are basically generated on demand and not really guaranteed to be stable over time.  Only make sense in the context of a single "Program"

File: pkg/engine/engine.go
Comment (g-linville) [reply pkg/engine/engine.go]: Fixed

File: pkg/credentials/credential.go
Comment (g-linville) [reply pkg/credentials/credential.go]: Fixed and renamed to `ToolName`. Though based on a discussion that I had with Taylor earlier today, we are probably going to need to use something else as the identifier/unique name for each credential

File: pkg/credentials/credential.go
Comment (g-linville) [inline pkg/credentials/credential.go:49]: Open for suggestions for something better than this. I need a way of combining toolName and credCtx such that it still creates a valid URL.

File: pkg/runner/runner.go
Comment (iwilltry42) [inline pkg/runner/runner.go:319]: Not sure if that'd be too much, but maybe we can have a flag for "reprompt" to update all cached credentials in a run? My laziness coming through.. :D

File: pkg/cli/gptscript.go
Comment (ibuildthecloud) [inline pkg/cli/gptscript.go:56]: Don't use `-c`.  Just make it long and verbose right now.  `-c` might be useful for something else more important later

File: pkg/cli/credential.go
Comment (ibuildthecloud) [inline pkg/cli/credential.go:21]: I've been using the Customize interface approach to change the default cobra.Command fields.  I feel like that is slightly less confusing pattern. I'm going to probably drop support for this style of passing it as a second arg.

https://github.com/gptscript-ai/gptscript/blob/main/pkg/cli/gptscript.go#L102

File: pkg/cli/credential.go
Comment (ibuildthecloud) [inline pkg/cli/credential.go:17]: Don't do `-A`.  Either have no short or if you really want a short value then `-a`.  The `-A` is very kubernetes-y.

File: pkg/runner/runner.go
Comment (g-linville) [reply pkg/runner/runner.go:319]: I asked Darren about it and he wants to hold off on this for now.

File: pkg/cli/credential_delete.go
Comment (g-linville) [reply pkg/cli/credential_delete.go:21]: For now, no.

File: pkg/cli/gptscript.go
Comment (g-linville) [reply pkg/cli/gptscript.go]: Fixed

File: pkg/cli/credential.go
Comment (g-linville) [reply pkg/cli/credential.go]: Fixed

File: pkg/cli/credential.go
Comment (g-linville) [reply pkg/cli/credential.go]: Fixed

File: docs/docs/03-tools/04-credentials.md
Comment (g-linville) [reply docs/docs/03-tools/04-credentials.md]: Fixed

**Review by iwilltry42 — APPROVED** [review]

**Comment by g-linville** [comment]
(I got verbal approval from Darren to merge.)

**Pull request merged**
